### PR TITLE
New Gamemode with cornucopia cards

### LINKED ISSUE #73: Alternative card deck: OWASP Cornucopia
Give the possibility to choose between different playing modes:
* EoP
* [OWASP Cornucopia](https://owasp.org/www-project-cornucopia/)

Cornucopia basically follows the same game flow as EoP, but might be better suited for threat modelling web applications.
These should just be UI changes (i.e. exchange images and make it configurable).

### PR DESCRIPTION
**Pull request by JonathanDeul**
implemented new gamemode with cornucopia cards. Also the starting suit can now be chaned depending on the gamemode.

Resolves #73

### CODE REVIEW

File: src/client/pages/create.js [hunk 1]
```diff
@@ -188,13 +198,22 @@ class Create extends React.Component {
...
+              </Col>
+            </FormGroup>
             <FormGroup row>
               <Label for="startSuit" sm={2}>Start Suit</Label>
               <Col sm={10}>
                 <Input type="select" name="startSuit" id="startSuit" onChange={e => this.onstartSuitUpdated(e)} value={this.state.startSuit}>
                   {
                     Object.keys(STARTING_CARD_MAP).map(suit => (
-                      <option value={suit} key={`start-suit-option-${suit}`}>{getTypeString(suit)}</option>
+                      <option value={suit} key={`start-suit-option-${suit}`}>{getTypeString((this.state.gamemode === "Cornucopia") ? "C" + suit : "" + suit)}</option>
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:216]: Maybe add the gameMode as a second parameter to getTypeString? I believe that might be better readable

File: src/utils/constants.js [hunk 2]
```diff
@@ -1,16 +1,15 @@
 export const DECK_HANDS = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'];
-export const DECK_SUITS = ['S', 'T', 'R', 'I', 'D', 'E'];
-export const INVALID_CARDS = ['T2', 'E2', 'E3', 'E4'];
-export const TRUMP_CARD_PREFIX = 'E';
-export const DEFAULT_START_SUIT = 'T'
-//TODO: Extract the following from INVALID_CARDS
+export const DECK_SUITS = ['A', 'B', 'C', 'D', 'E', 'T'];
+export const INVALID_CARDS = ['E2', 'T2', 'T3', 'T4'];
```
Comment (ChristophNiehoff) [inline src/utils/constants.js:3]: Why is the first card E2 and not B2?

File: src/client/components/dealtcard/dealtcard.js [hunk 3]
```diff
@@ -4,12 +4,13 @@ import PropTypes from 'prop-types';
...
   static propTypes = {
     card: PropTypes.string.isRequired,
+    gamemode: PropTypes.string.isRequired,
   };
 
   render() {
     const cardClass = this.props.card ? this.props.card.toLowerCase() : '-none';
     return (
-      <div className={`playing-card card${cardClass} active card-rounded scaled-big`} />
+      <div className={`playing-card ${(this.props.gamemode === "Cornucopia") ? "c" : ""}card${this.props.card.toLowerCase()} active card-rounded scaled-big`} />
```
Comment (ChristophNiehoff) [inline src/client/components/dealtcard/dealtcard.js:13]: Very often you are using gameMode === Cornucopia. Could you move this to a dedicated function like isGameModeCornucopia?

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]
Already works pretty well!

In the leader board the abbreviation of the played card is very cryptic. For EoP (i.e. for STRIDE), I'd like to have how it was before. For Cornucopia it is more difficult, as several suits have the same starting letter. I propose using the following abbreviations:
* AuthZ
* Crypt
* Sessn
* AuthN
* Data
* Cornu
I hope there enough space for that.

When entering a threat the player can choose a stride categorie. How is this done for Cornucopia?

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]

**Comment by ChristophNiehoff** [comment]
@JonathanDeul  I really like to modified cards!
One thing: For some of the cards (especially AuthZ) there is a white line on the left. Seems to be some kind of offset. Is this something that can easily be fixed?  If not, then it's not important.
whiteline

**Comment by JonathanDeul** [comment]
> @JonathanDeul I really like the modified cards!
> One thing: For some of the cards (especially AuthZ) there is a white line on the left. Seems to be some kind of offset. Is this something that can easily be fixed? If not, then it's not important.
> whiteline

yep. I will fix that!

File: src/client/components/sidebar/sidebar.js [hunk 4]
```diff
@@ -45,17 +45,16 @@ class Sidebar extends React.Component {
...
         <Button color={(isLastToPass) ? "warning" : "secondary"} className="pass" size="lg" block disabled={
-
           this.props.ctx.phase !== "threats" ||
           this.props.G.passed.includes(this.props.playerID) ||
           !this.props.active
         } onClick={() => { this.props.moves.pass() }}>
           Pass
         </Button>
-        <DealtCard card={dealtCard} />
+        <DealtCard card={dealtCard} gameMode={currentGameMode} />
```
Comment (ChristophNiehoff) [inline src/client/components/sidebar/sidebar.js:57]: Is there a reason why you cannot directly use `this.props.G.gameMode` at this place?

File: src/client/components/threatbar/threatbar.js [hunk 5]
```diff
@@ -243,48 +246,47 @@ class Threatbar extends React.Component {
...
-                  <option value="S">Spoofing</option>
-                  <option value="T">Tampering</option>
-                  <option value="R">Repudiation</option>
-                  <option value="I">Information Disclosure</option>
-                  <option value="D">Denial of Service</option>
-                  <option value="E">Elevation of Privilege</option>
+                <Input type="select" name="type" id="type" disabled={this.props.G.threat.owner !== this.props.playerID} value={this.props.G.threat.type} onChange={(e) => this.props.moves.updateThreat("type", e.target.value)}>
+                  {
+                    Object.keys(STARTING_CARD_MAP).map(suit => (
+                      <option value={suit}>{getTypeString(suit, this.props.gameMode)}</option>
```
Comment (ChristophNiehoff) [inline src/client/components/threatbar/threatbar.js:256]: In react, when a component is added through a loop, one should add a `key`, see [the docs](https://reactjs.org/docs/lists-and-keys.html#keys).

So, you should add a property like 
```js
key={`threat-category-${suit}`}
```

File: src/client/pages/create.js [hunk 6]
```diff
@@ -33,13 +33,15 @@ class Create extends React.Component {
       creating: false,
       created: false,
       model: null,
-      startSuit: "T",
+      startSuit: "E",
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:36]: Could you use `DEFAULT_START_SUIT` here?

File: src/client/pages/create.js [hunk 7]
```diff
@@ -33,13 +33,15 @@ class Create extends React.Component {
...
       provideModelThruAlternativeChannel: false,
+      gameMode: "Elevation of Privilege",
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:38]: Introduce a config parameter like `DEFAULT_GAME_MODE`?

File: src/utils/constants.js [hunk 8]
```diff
@@ -1,16 +1,15 @@
 export const DECK_HANDS = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'];
-export const DECK_SUITS = ['S', 'T', 'R', 'I', 'D', 'E'];
-export const INVALID_CARDS = ['T2', 'E2', 'E3', 'E4'];
-export const TRUMP_CARD_PREFIX = 'E';
-export const DEFAULT_START_SUIT = 'T'
-//TODO: Extract the following from INVALID_CARDS
+export const DECK_SUITS = ['A', 'B', 'C', 'D', 'E', 'T'];
+export const INVALID_CARDS = ['B2', 'T2', 'T3', 'T4'];
```
Comment (ChristophNiehoff) [inline src/utils/constants.js:3]: Are these cards also excluded from Cornucopia?Or do we need separate arrays for EoP and Cornucopia?

File: src/utils/utils.js [hunk 9]
```diff
@@ -1,25 +1,37 @@
...
-  for (let i=0; i<dealt.length; i++) {
+  for (let i = 0; i < dealt.length; i++) {
     let idx = parseInt(order[i]);
     cards[idx] = dealt[i];
   }
   return cards;
 }
 
+export function isGameModeCornucopia(gameMode) {
+  return (gameMode === "OWASP Cornucopia");
```
Comment (ChristophNiehoff) [inline src/utils/utils.js:19]: Maybe it's a good idea to introduce constants like
```js
const GAMEMODE_EOP = "Elevation of Privilege";
const GAMEMODE_CORNUCOPIA = "OWASP Cornucopia";
```
in `constants.js`

Then one could use them consistently throughout the code (i.e. in the dopdowns, in this function etc.), to nt have these magic strings.

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]

File: cornucopiaCards/package.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+  "name": "cornucopia-cards",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "CC BY 3.0"
```
Comment (ChristophNiehoff) [inline cornucopiaCards/package.json:11]: This should be the shareAlike license.  CC-BY-SA-3.0

File: cornucopiaCards/package.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "cornucopia-cards",
+  "version": "1.0.0",
+  "description": "",
```
Comment (ChristophNiehoff) [inline cornucopiaCards/package.json:4]: We need a proper description and maybe remove the "main"-field, as this file is not present

File: cornucopiaCards/package.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "cornucopia-cards",
```
Comment (ChristophNiehoff) [inline cornucopiaCards/package.json:2]: Maybe even call the package "cornucopia-cards-modified"?

File: cornucopiaCards/package.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
```
Comment (ChristophNiehoff) [inline cornucopiaCards/package.json:10]: Author should probably be "TNG Technology Consulting"

File: src/client/components/deck/deck.js [hunk 14]
```diff
@@ -1,6 +1,8 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { getValidMoves } from '../../../utils/utils';
+import { getValidMoves, isGameModeCornucopia } from '../../../utils/utils';
+import { render } from 'react-dom';
```
Comment (ChristophNiehoff) [inline src/client/components/deck/deck.js:4]: Do you need this import? Or is it an error?

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]
We still need a license file for the new subfolder. This should probably say that we modified the original cards and should state the original authors. Maybe we also have to include the CC-BY-SA-3.0 license text (depending on what the licedse actually says...)

And we also have to modify the toplevel license file such that it explicitly excludes the new subfolder.

**Pull request merged**